lib/prelude.dx

```diff
@@ -260,16 +309,29 @@ def withReader
       : {|eff} a =
     runReader init action
 
+def MonoidLifter (b:Type) (w:Type) : Type = h:Type -> AccumMonoid h b ?=> AccumMonoid h w
+
 def runAccum
-      (action: (h:Type ?-> Ref h w -> {Accum h|eff} a))
+      (mlift:MonoidLifter b w) ?=>
+      (bm:Monoid b)
+      (action: (h:Type ?-> AccumMonoid h b ?=> Ref h w -> {Accum h|eff} a))
       : {|eff} (a & w) =
-    def explicitAction (h':Type) (ref:Ref h' w) : {Accum h'|eff} a = action ref
-    %runWriter explicitAction
+    -- Normally, only the ?=> lambda binders participate in dictionary synthesis,
+    -- so we need to explicitly declare `m` as a hint.
+    %instance bmHint = bm
+    empty : b = mempty
+    combine : b -> b -> b = mcombine
+    def explicitAction (h':Type) (ref:Ref h' w) : {Accum h'|eff} a =
+      %instance accumBaseMonoidHint : AccumMonoid h' b = UnsafeMkAccumMonoid bm
+      action ref
+    %runWriter empty combine explicitAction
```

Comment:
After reading the rest of this PR and coming back to this, I'm trying to figure out how the dictionary synthesis works.

1. We ask for a `mlift:MonoidLifter b w` dictionary, which isn't an ADT but is actually a *function* `AccumMonoid h b ?=> AccumMonoid h w`. The compiler thus assumes that the caller of this function is going to tell us how to convert from a monoid in `b` to a monoid in `w`.
2. We build an "explicit action". Inside that action, we introduce a local instance `bmHint` for the base monoid. That instance is then passed to `action` implicitly.
3. We run the action.

I don't understand what happens to `mlift`. It seems like it is introduced to a local scope and thus would participate in dictionary synthesis if we asked for a `AccumMonoid h w` instance. But we don't seem to actually do this here? The action is written so that it only requires a monoid instance for the base type, so we wouldn't ever need to synthesize `AccumMonoid h w` in this function. (If anything, it would be synthesized inside `action`, but then `mlift` won't be in scope anymore, so I don't see how it participates?)

Reply:
Hm, that seems scary! Maybe deserves a bunch of warning comments to make sure it's respected?

Wait, does that mean that `tableAccumMonoid` could just be
```
@instance
def tableAccumMonoid ((UnsafeMkAccumMonoid m):AccumMonoid h w) ?=> : AccumMonoid h (n=>w) = todo
```
because the implementation isn't used?

Reply:
I think it's used to provide the `mempty` initialization to `runAccum`. Yes, probably deserves some warning comments. I think that was the intention behind calling it `UnsafeMkAccumMonoid`.

Reply:
Doesn't the `mempty` initialization come from `bm: Monoid b`?

Reply:
Anyways, I have strong feelings about code that looks like it's doing something important, but isn't actually used. If this is actually just a case of the compiler doing what it wants and the userspace being safety rails to get a better error message, I'd much prefer something like
```
data AccumMonoid h w = UnsafePromiseAccumMonoid
```
that holds no data and obviously isn't used to actually do anything.

Reply:
Yep, on closer examination I agree it's not used at all. I think the reason that Adam decided to do it that way (rather than making the internal `runWriter` closer to the prelude version, taking the lifter as an argument) is that we need to be able to synthesize zeros when we parallelize accumulations, and we don't have access to the lifter there, so it made sense to make it a built-in. But I should stop speculating and wait for Adam's take tomorrow.

Reply:
Oh, I should add: even though `mlift` isn't used in this function, the payload of an `AccumMonoid` is used in `(+=)`, so it does need an implementation.

Reply:
Ah, you're totally right, I didn't catch that!

I wonder, then, whether the `mlift:MonoidLifter b w` does anything? If `(+=)` requires `AccumMonoid h w`, I think that means we need to derive `AccumMonoid h w` from `AccumMonoid h b` in order to use `(+=)` inside `action`. So maybe there's no need to prove that separately as a safety check since it's already derived when building `action`?